nntrainer/layers/cl_layers/concat_cl.cpp

```diff
@@ -233,12 +233,54 @@ std::string concat_cl_axis1_kernel_ =
 })";
 
 namespace nntrainer {
-ConcatLayerCl::ConcatLayerCl() : Layer() {}
+ConcatLayerCl::ConcatLayerCl() : LayerImplCl() {}
 
 static constexpr size_t SINGLE_INOUT_IDX = 0;
 static constexpr size_t INPUT_IDX_1 = 0;
 static constexpr size_t INPUT_IDX_2 = 1;
 
+bool ConcatLayerCl::registerClKernels() {
+  ClContext::SharedPtrClKernel kernel_concat_ptr = nullptr;
+
+  kernel_concat_ptr =
+    cl_context_ref.registerClKernel(concat_cl_axis1_kernel_, "concat_cl_axis1");
+  NNTR_THROW_IF(!kernel_concat_ptr, std::runtime_error)
+    << "OpenCL Error: Fail to register concat_cl_axis1 kernel";
+  layer_kernel_ptrs.emplace_back(kernel_concat_ptr);
+
+  kernel_concat_ptr =
+    cl_context_ref.registerClKernel(concat_cl_axis2_kernel_, "concat_cl_axis2");
+  NNTR_THROW_IF(!kernel_concat_ptr, std::runtime_error)
+    << "OpenCL Error: Fail to register concat_cl_axis2 kernel";
+  layer_kernel_ptrs.emplace_back(kernel_concat_ptr);
+
+  kernel_concat_ptr =
+    cl_context_ref.registerClKernel(concat_cl_axis3_kernel_, "concat_cl_axis3");
+  NNTR_THROW_IF(!kernel_concat_ptr, std::runtime_error)
+    << "OpenCL Error: Fail to register concat_cl_axis3 kernel";
+  layer_kernel_ptrs.emplace_back(kernel_concat_ptr);
+
+  kernel_concat_ptr = cl_context_ref.registerClKernel(
+    concat_cl_axis1_kernel_fp16_, "concat_cl_axis1_fp16");
+  NNTR_THROW_IF(!kernel_concat_ptr, std::runtime_error)
+    << "OpenCL Error: Fail to register concat_cl_axis1_fp16 kernel";
+  layer_kernel_ptrs.emplace_back(kernel_concat_ptr);
+
+  kernel_concat_ptr = cl_context_ref.registerClKernel(
+    concat_cl_axis2_kernel_fp16_, "concat_cl_axis2_fp16");
+  NNTR_THROW_IF(!kernel_concat_ptr, std::runtime_error)
+    << "OpenCL Error: Fail to register concat_cl_axis2_fp16 kernel";
+  layer_kernel_ptrs.emplace_back(kernel_concat_ptr);
+
+  kernel_concat_ptr = cl_context_ref.registerClKernel(
+    concat_cl_axis3_kernel_fp16_, "concat_cl_axis3_fp16");
+  NNTR_THROW_IF(!kernel_concat_ptr, std::runtime_error)
+    << "OpenCL Error: Fail to register concat_cl_axis3_fp16 kernel";
+  layer_kernel_ptrs.emplace_back(kernel_concat_ptr);
+
+  return true;
```

Comment:
quick question! can't ConcatLayerCl::registerClKernels() be called twice?
assume it is called for a second time, would it throw a runtime error or return true?

Reply:
I assumed it is only called in `add_default_object` only, which is called in `registerer` ; the `registerer` is called only once. However, it seems better to check it. I will update it. 

```
ClContext &ClContext::Global() {
  static ClContext instance;

  // initializing commandqueue and context
  bool result = instance.clInit();

  if (!result) {
    ml_loge("cl_context: opencl command queue creation failed");
  }

  /// in g++ there is a bug that hangs up if caller throws,
  /// so registerer is noexcept although it'd better not
  /// https://gcc.gnu.org/bugzilla/show_bug.cgi?id=70298
  std::call_once(global_cl_context_init_flag, registerer, std::ref(instance));
  return instance;
}
```

Reply:
I added a condition to check it in the `registerClKernels()` as well.
https://github.com/nnstreamer/nntrainer/blob/6dc250f57117375d5d55a981976f59cdf4e8540c/nntrainer/layers/cl_layers/reshape_cl.cpp#L55